# oviIN
There is 1 oviIN neuron in each hemisphere of the fly brain.

In [28]:
from neuprint import Client
# remove my token before making notebook public
c = Client('neuprint.janelia.org', dataset='hemibrain:v1.2.1', token='eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImdnMjExNEBjb2x1bWJpYS5lZHUiLCJsZXZlbCI6Im5vYXV0aCIsImltYWdlLXVybCI6Imh0dHBzOi8vbGgzLmdvb2dsZXVzZXJjb250ZW50LmNvbS9hLS9BT2gxNEdpb1lJLUVPLWdidGxPRTh6SmQ0eF9ZQ1Y4ZHF0YVFjWGlHeG5CMz1zOTYtYz9zej01MD9zej01MCIsImV4cCI6MTgxMDUyOTYzNH0.jv9eR0SH5RhfBdXrtp4r-dDFOhcsT8GBbE4v69ysCKs') 
c.fetch_version()

# import important stuff
import numpy as np
import pandas as pd

In [3]:
# body IDs of oviINs from Neuprint
oviINr_bodyID = 423101189
oviINl_bodyID = 485934965

Fetching some basic info about the oviINs. This info is somewhat useless to me. Especially because the Neuprint documentation doesn't go into detail about what data is in some of these columns.

In [5]:
from neuprint import fetch_neurons

R_ovi_neuron_df, roi_counts_df = fetch_neurons([oviINr_bodyID,oviINl_bodyID])

In [6]:
R_ovi_neuron_df

bodyId instance   type   pre   post  downstream  upstream  mito  \
0  423101189  oviIN_R  oviIN  6863  23029       60603     23029  2472   
1  485934965  oviIN_L  oviIN  6542  15998       58310     15998  2208   

          size  status  cropped     statusLabel cellBodyFiber somaRadius  \
0  10033593370  Traced    False  Roughly traced          None       None   
1   8493821787  Traced    False  Roughly traced          None       None   

  somaLocation                                            roiInfo notes  \
0         None  {'SNP(R)': {'pre': 4773, 'post': 13645, 'downs...  None   
1         None  {'SNP(L)': {'pre': 4184, 'post': 10529, 'downs...  None   

                                           inputRois  \
0  [ATL(R), CAN(R), CRE(-ROB,-RUB)(R), CRE(-RUB)(...   
1  [ATL(L), CRE(-ROB,-RUB)(R), CRE(-RUB)(L), CRE(...   

                                          outputRois  
0  [CAN(R), CRE(-ROB,-RUB)(R), CRE(-RUB)(L), CRE(...  
1  [ATL(L), CRE(-ROB,-RUB)(R), CRE(-RUB)(L), CRE(...

In [7]:
roi_counts_df

bodyId                roi   pre   post  downstream  upstream  mito
0   423101189             ATL(R)     0     18           0        18     2
1   423101189             CAN(R)     1     24           3        24     5
2   423101189  CRE(-ROB,-RUB)(R)   420   3710        3414      3710   286
3   423101189       CRE(-RUB)(L)    18      9         184         9     2
4   423101189             CRE(L)    18      9         184         9     2
5   423101189             CRE(R)   420   3715        3414      3715   286
6   423101189             FLA(R)     0     22           0        22     3
7   423101189             GOR(R)     4     23          19        23     6
8   423101189                 IB    10    231          47       231    27
9   423101189                INP   448   3973        3645      3973   317
10  423101189        LAL(-GA)(R)    16   1775         106      1775   112
11  423101189             LAL(R)    16   1775         106      1775   112
12  423101189              LX(R)    16   1775         106      1775   112
13  423101189         NotPrimary     1     13           1        13   102
14  423101189               PENP     1     46           3        46     8
15  423101189             ROB(R)     0      5           0         5     0
16  423101189             SIP(L)    17     23         160        23     3
17  423101189             SIP(R)   207    479        1985       479    54
18  423101189             SMP(L)  1572   1563       14389      1563   314
19  423101189             SMP(R)  4566  13166       40140     13166  1386
20  423101189             SNP(L)  1589   1586       14549      1586   317
21  423101189             SNP(R)  4773  13645       42125     13645  1440
22  423101189             SPS(R)     5    173          19       173    24
23  423101189             VES(R)    26   1795         136      1795   146
24  423101189               VMNP    35   1991         174      1991   176
25  485934965             ATL(L)     4     11          30        11     3
26  485934965  CRE(-ROB,-RUB)(R)     7      2          55         2     1
27  485934965       CRE(-RUB)(L)   378   2204        3166      2204   275
28  485934965             CRE(L)   378   2204        3166      2204   275
29  485934965             CRE(R)     7      2          55         2     1
30  485934965             GOR(L)     4     65          21        65    10
31  485934965                 IB     4     47          35        47     4
32  485934965                INP   397   2266        3316      2266   284
33  485934965             LAL(L)     7    795          35       795    81
34  485934965              LX(L)     7    795          35       795    81
35  485934965         NotPrimary     9     20          25        20    99
36  485934965             SCL(L)     4      2          30         2     1
37  485934965             SIP(L)   128    171        1102       171    27
38  485934965             SIP(R)    18     13         168        13     4
39  485934965             SMP(L)  4056  10358       35826     10358  1316
40  485934965             SMP(R)  1916   2024       17783      2024   355
41  485934965             SNP(L)  4184  10529       36928     10529  1343
42  485934965             SNP(R)  1934   2037       17951      2037   359
43  485934965             VES(L)     7    286          34       286    32
44  485934965               VMNP    11    351          55       351    42

- How many synaptic partners does oviIN have? pre? post? total?
- How many synaptic sites? pre? post? get descriptive
- Where does ovi rank in terms of number of synaptic partners? in terms of total numbers of synaptic sites?
- What is the average number of neuropils that any given neuron connects to and how does oviIN rank?

## oviIN's personal specs

In [5]:
# inputs to oviINr
from neuprint import fetch_simple_connections
ovi_inputs = fetch_simple_connections(None,oviINr_bodyID)
ovi_inputs

bodyId_pre  bodyId_post  weight type_pre type_post instance_pre  \
0      611015907    423101189     287   SMP176     oviIN     SMP176_R   
1     5813087891    423101189     255   SMP544     oviIN     SMP544_R   
2     1598011458    423101189     243   LAL134     oviIN     LAL134_R   
3     5813022375    423101189     236    IB017     oviIN      IB017_R   
4      448342846    423101189     225   CRE075     oviIN     CRE075_R   
...          ...          ...     ...      ...       ...          ...   
2519  5901221890    423101189       1   SLP422     oviIN     SLP422_R   
2520  5901226036    423101189       1     None     oviIN         None   
2521  5901230025    423101189       1     None     oviIN         None   
2522  5901231318    423101189       1     None     oviIN         None   
2523  7112624972    423101189       1     None     oviIN         None   

     instance_post                                       conn_roiInfo  
0          oviIN_R  {'INP': {'pre': 54, 'post': 53}, 'CRE(R)': {'p...  
1          oviIN_R  {'LX(R)': {'pre': 66, 'post': 65}, 'LAL(R)': {...  
2          oviIN_R  {'LX(R)': {'pre': 42, 'post': 42}, 'LAL(R)': {...  
3          oviIN_R  {'SNP(R)': {'pre': 97, 'post': 97}, 'SMP(R)': ...  
4          oviIN_R  {'INP': {'pre': 199, 'post': 199}, 'CRE(R)': {...  
...            ...                                                ...  
2519       oviIN_R  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...  
2520       oviIN_R  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...  
2521       oviIN_R  {'VMNP': {'pre': 1, 'post': 1}, 'VES(R)': {'pr...  
2522       oviIN_R  {'SNP(L)': {'pre': 1, 'post': 1}, 'SMP(L)': {'...  
2523       oviIN_R  {'INP': {'pre': 1, 'post': 1}, 'CRE(R)': {'pre...  

[2524 rows x 8 columns]

The number of unique bodyIDs pre-synaptic to oviINr.

In [8]:
ovi_inputs['bodyId_pre'].nunique()

2524

The number of total synaptic sites from pre-synaptic neurons onto oviINr. This number is slightly lower than the number given by the Neuprint GUI (23029).

In [9]:
ovi_inputs['weight'].sum()

21624

Interestingly and thankfully, fetch_adjacencies gives the same results when include_nonprimary=False. 

In [18]:
from neuprint import fetch_adjacencies
n_df, conn_df = fetch_adjacencies(None,oviINr_bodyID)

In [19]:
conn_df

bodyId_pre  bodyId_post     roi  weight
0      234630133    423101189  CRE(R)       2
1      263674097    423101189  SMP(R)       2
2      266187480    423101189  SMP(R)       1
3      266187559    423101189  SMP(R)       3
4      267214250    423101189  SMP(R)       9
...          ...          ...     ...     ...
3526  6400000773    423101189  SMP(R)       2
3527  7112622044    423101189  SIP(R)       1
3528  7112622044    423101189  SMP(R)       1
3529  7112622044    423101189  SMP(L)       1
3530  7112624972    423101189  CRE(R)       1

[3525 rows x 4 columns]

In [20]:
conn_df['bodyId_pre'].nunique()

2524

In [21]:
conn_df['weight'].sum()

21624

In [22]:
# outputs from oviINr
from neuprint import fetch_simple_connections
ovi_outputs = fetch_simple_connections(oviINr_bodyID,None)
ovi_outputs

bodyId_pre  bodyId_post  weight type_pre type_post instance_pre  \
0      423101189   5813021602     320    oviIN    SMP198      oviIN_R   
1      423101189    452689494     239    oviIN    SMP550      oviIN_R   
2      423101189    611629428     238    oviIN    CRE077      oviIN_R   
3      423101189    542376096     208    oviIN    SMP237      oviIN_R   
4      423101189   5813021005     199    oviIN    SMP507      oviIN_R   
...          ...          ...     ...      ...       ...          ...   
3436   423101189   5901197274       1    oviIN    SMP528      oviIN_R   
3437   423101189   5901203310       1    oviIN      None      oviIN_R   
3438   423101189   5901220525       1    oviIN      None      oviIN_R   
3439   423101189   7112625063       1    oviIN      None      oviIN_R   
3440   423101189   7112625282       1    oviIN      None      oviIN_R   

        instance_post                                       conn_roiInfo  
0            SMP198_R  {'SNP(R)': {'pre': 276, 'post': 276}, 'SMP(R)'...  
1            SMP550_R  {'SNP(R)': {'pre': 239, 'post': 239}, 'SMP(R)'...  
2            CRE077_R  {'SNP(R)': {'pre': 216, 'post': 216}, 'SMP(R)'...  
3            SMP237_R  {'SNP(R)': {'pre': 203, 'post': 203}, 'SMP(R)'...  
4            SMP507_R  {'SNP(R)': {'pre': 199, 'post': 199}, 'SMP(R)'...  
...               ...                                                ...  
3436  SMP528(PDM24)_L  {'SNP(L)': {'pre': 1, 'post': 1}, 'SMP(L)': {'...  
3437        (PVM12)_L  {'SNP(L)': {'pre': 1, 'post': 1}, 'SMP(L)': {'...  
3438             None  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...  
3439             None  {'SNP(R)': {'pre': 1, 'post': 1}, 'SMP(R)': {'...  
3440             None  {'SNP(L)': {'pre': 1, 'post': 1}, 'SMP(L)': {'...  

[3441 rows x 8 columns]

The number of unique bodyIDs post-synaptic to oviINr

In [23]:
ovi_outputs['bodyId_post'].nunique()

3441

The number of total post-synaptic sites from oviINr. This number is much higher than that given in Neuprint (6863). Somehow, they must've filtered something out. Maybe I should disregard the GUI because nothing is really making sense there...

In [24]:
ovi_outputs['weight'].sum()

29274

The total number of direct partners for oviINr.

In [32]:
ovi_partners = pd.concat([ovi_inputs['bodyId_pre'],ovi_outputs['bodyId_post']],ignore_index=True)
ovi_partners.nunique()

4548

## Where the synapses are
Getting descriptive about oviIN's synaptic sites. Not only how many, but where they are located.

## oviIN in comparison to other neurons in the fly brain
Where oviIN ranks in terms of number of synpatic partners, number of synaptic sites (pre and post), and number of neuropils it connects to.

In [37]:
from neuprint import fetch_primary_rois
primROIs = fetch_primary_rois()

In [54]:
from neuprint import fetch_neurons
from neuprint import NeuronCriteria as NC

# these all time out...
#all_the_neurons, roi_counts_df = fetch_neurons(NC(inputRois=['PB','AB(L)']))
all_the_neurons, roi_counts_df = fetch_neurons(NC(rois=primROIs))
all_the_neurons

Empty DataFrame
Columns: [bodyId, instance, type, pre, post, downstream, upstream, mito, size, status, cropped, statusLabel, cellBodyFiber, somaRadius, somaLocation, roiInfo, notes, inputRois, outputRois]
Index: []

In [49]:
primROIs

['AB(L)',
 'AB(R)',
 'AL(L)',
 'AL(R)',
 'AME(R)',
 'AOTU(R)',
 'ATL(L)',
 'ATL(R)',
 'AVLP(R)',
 'BU(L)',
 'BU(R)',
 'CA(L)',
 'CA(R)',
 'CAN(R)',
 'CRE(L)',
 'CRE(R)',
 'EB',
 'EPA(L)',
 'EPA(R)',
 'FB',
 'FLA(R)',
 'GNG',
 'GOR(L)',
 'GOR(R)',
 'IB',
 'ICL(L)',
 'ICL(R)',
 'IPS(R)',
 'LAL(L)',
 'LAL(R)',
 'LH(R)',
 'LO(R)',
 'LOP(R)',
 'ME(R)',
 'NO',
 'PB',
 'PED(R)',
 'PLP(R)',
 'PRW',
 'PVLP(R)',
 'SAD',
 'SCL(L)',
 'SCL(R)',
 'SIP(L)',
 'SIP(R)',
 'SLP(R)',
 'SMP(L)',
 'SMP(R)',
 'SPS(L)',
 'SPS(R)',
 'VES(L)',
 'VES(R)',
 'WED(R)',
 "a'L(L)",
 "a'L(R)",
 'aL(L)',
 'aL(R)',
 "b'L(L)",
 "b'L(R)",
 'bL(L)',
 'bL(R)',
 'gL(L)',
 'gL(R)']